Network stack and power management

tbursztyka opened the pull request:
Adding net_if awareness of PM state.

Commit:
net: Add preliminaly support for suspending/resuming a net interface

Such state needs to be set _from_ the PM API functions and not the other
way round. So if a network device driver does not support such API, it
will not be able to set the core net_if on PM state, obviously.

Currently, these functions only set/unset NET_IF_SUSPENDED flag.

More logic will be added later, to decide whether the net_if can be
actually set to suspend mode or not and also to take care of all timers
related to the interface.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net: Enable PM settings on network devices

By changing the various *NET_DEVICE* macros. It is up to the device
drivers to either set a proper PM function or, if not supported or PM
disabled, to use device_pm_control_nop relevantly.

All existing macro calls are updated. Since no PM support was added so
far, device_pm_control_nop is used as the default everywhere.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
device: Provide a dummy macro for device_pm_control_nop on disabled PM

This will prove to be useful on enabling PM setting for the network
devices.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/clock_gate: Support on/off operationf in mcux's SIM driver

Logic to enable/disable the relevant clock was fully missing.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/arm: Set the clock source for the ethernet port on NXP's k6x

SIM core system clock is being used, but more importantly this will
enable to get the SIM clock controller in use for power management
purposes in MCUX ethernet driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ethernet: Add support for power management into MCUX

Requesting the clock controller to switch on or off the device.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net: Expose a way to know if the network interface is suspended or not

And use it in the shell to display such information then.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/shell: Add a quick way to suspend/resume a network interface

Meaningful for testing purpose only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
samples/net: Adding relevant parts to test PM on echo server

Just to get something to test for PM, via frdm_k64f board. So only this
board will get PM enabled.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
power management: Add k6x SOC core list

Just inserting mcux's ethernet device there.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@MaureenHelm Could you, or one of your colleague, check on 
- drivers/clock_gate: Support on/off operationf in mcux's SIM driver 
- drivers/ethernet: Add support for power management into MCUX 

It works and does not work: the ethernet port does not seem to shut down fully (so the clock is not off? wrong driver or?). Also, using ENET_Denitit() generates a hw crash.

pabigot commented on subsys/power/device.c:
```diff
@@ -29,6 +29,15 @@ static const char *const core_devices[] = {
+#if defined(CONFIG_SOC_SERIES_KINETIS_K6X)
+#define MAX_PM_DEVICES		1
+#define MAX_DEV_NAME_LEN	16
+static const char core_devices[][MAX_DEV_NAME_LEN] = {
```
I've seen this pattern before; is there any reason why `MAX_DEV_NAME_LEN` must be reserved for each entry?  This could be:
```
static const char *core_devices[] = {
    DT_ETH_MCUX_0_NAME,
}
```
which would save a few bytes and allow string constant sharing.

tbursztyka replied:
That's how this PM awfulness was made. It's not the point of this PR to change anything there.

tbursztyka replied:
wait, that was changed recently, that's a rebasing left-over.

jukkar committed:
net: if: Add function to check any pending TX packets

This function can be used for example by network power management
to check if the network interface can be suspended or not.
If there are network packets in transmit queue, then the network
interface cannot be suspended yet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: tc: Return information whether TX pkt was queued

We need to know whether the net_pkt was successfully placed
to transmit queue. It is possible in TX side, that the net_pkt
is already in TX queue when for example TCP packet is
re-transmitted, in which case the queue submit will fail.
This cannot happen in RX side as there are no timers involved.

It is required to check about such pending flag before trying to submit
it into the queue. Indeed, the work queue could be scheduled right after
such queuing, thus checking for the pending flag afterwards would
provide a false information.

It is unfortunate k_work_submit_to_queue() does not return anything as
it would simplify the code then.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jukkar commented on include/net/net_if.h:
```diff
@@ -2101,6 +2109,27 @@ void net_if_unset_promisc(struct net_if *iface);
+ *
+ * @param iface Pointer to network interface
+ *
+ * @return True if there are no any pending TX network packets for this network
```
I just realized that the text here went wrong. The function returns true if there are pending packets, and false otherwise, now the description here says otherwise. @tbursztyka can you tweak this

tbursztyka replied:
indeed, will do

jukkar reviewed (approved):
LGTM

dleach02 commented:
> @MaureenHelm Could you, or one of your colleague, check on
> 
> * drivers/clock_gate: Support on/off operationf in mcux's SIM driver
> * drivers/ethernet: Add support for power management into MCUX
> 
> It works and does not work: the ethernet port does not seem to shut down fully (so the clock is not off? wrong driver or?). Also, using ENET_Denitit() generates a hw crash.

@tbursztyka, what do you check to determine that the port is not shut down fully?

tbursztyka commented on samples/net/sockets/echo_server/prj.conf:
```diff
@@ -50,3 +50,7 @@ CONFIG_POSIX_MAX_FDS=12
+
+
+CONFIG_NET_DHCPV4=y
+CONFIG_ETHERNET_LOG_LEVEL_DBG=y
```
no, looks like I forgot to send the updated version since I tested with Jukka's patch

tbursztyka commented:
@dleach02 I assume the port should not have any lights on. But it's not the case, it is still on. So could you verify if the clock_control driver is right (according to datasheet it is the sim clock gating controller that works for ethernet.). And why Enet_deinit() crashes, that's a nxp's hal issue

dleach02 commented on drivers/clock_control/clock_control_mcux_sim.c:
```diff
@@ -15,11 +15,19 @@ LOG_MODULE_REGISTER(clock_control);
 static int mcux_sim_on(struct device *dev, clock_control_subsys_t sub_system)
 {
+	clock_ip_name_t clock_ip_name = (clock_name_t) sub_system;
+
```
The cast of sub_system is incorrect. The sub_system value is actually a clock_ip_name_t which is 32bits in size. The clock_name_t is only 16 bits in size so the cast causes the value to be zero.

dleach02 commented on drivers/clock_control/clock_control_mcux_sim.c:
```diff
@@ -15,11 +15,19 @@ LOG_MODULE_REGISTER(clock_control);
 static int mcux_sim_off(struct device *dev, clock_control_subsys_t sub_system)
 {
+	clock_ip_name_t clock_ip_name = (clock_name_t) sub_system;
+
```
same as comment on mcux_sim_on()

dleach02 reviewed (commented):
@tbursztyka, the yellow LED appears to be tied directly to power if I'm reading the schematics correctly.

Also, I was able to reproduce the crash in ENET_Deinit() but then it stopped crashing??? I'm not sure why it stopped because I don't believe I did anything different from when it crashed to it crashing. 

Can you first do the fixes I identified for the clock on/off and then redo the code to use ENET_Deinit() to see if you still get the crashes? I also looked at some internal test case and they call ENET_Reset() before the Deinit(). Not sure that really matters though.

tbursztyka commented:
@dleach02 same here, retested with Enet_deinit() and it worked... ok then.

dleach02 commented:
@tbursztyka, I believe the fault has something to do with accessing the register at address 0x400c0000 when the ENET is disabled. In the case of when it was happening to both of us the first time ENET_Deinit was called I think the fault was actually from an interrupt handler accessing the register. This morning I was able to reproduce it by doing the 'net suspend 1' twice in a row and now the fault moved to <eth_mcux_phy_enter_reset+22> which I suspect the compiler inlined ENET_StartSMIWrite and there is a write to base->MMFR. 

I don't know why this is as I'm not familiar with the ENET HW. Still thinking about this.

dleach02 commented:
@tbursztyka, @MaureenHelm 

Tomasz, the baseline problem is that we have code floating around that is accessing the  ENET registers after the code has turned off the clocks to this module. This results in the crashes we see. You can reproduce this directly by doing back to back 'enet suspend 1' where the second one will crash. The module won't respond causing a bus error. You can sort of see this with GDB as well by doing an 'x/xw 0x400c0004' before and after doing the ENET_Deinit()

What I appears to be happening when we see the crash you initially reported with the ENET_Deinit() is that an ENET ISR is occurring (might be the error ISR) and the code in the ISR tries to read the ENET register. 

This implies that if you want to power down the ethernet you are going to need to ensure that there is no more interrupts that can come from the controller. We might need to add in an ENET_Reset() call before doing the deinit(). You also need to add a check in your logic to ensure that the ENET is not already in suspend state to avoid trying to put the device in suspend again (and crashing).

dleach02 commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -166,6 +177,64 @@ rx_buffer[CONFIG_ETH_MCUX_RX_BUFFERS][ETH_MCUX_BUFFER_SIZE];
+
+			ret = net_if_suspend(eth_ctx->iface);
+			if (ret == -EBUSY) {
+				goto out;
```
I would change this test of ret to:

if (!ret) 

This will prevent the code from running twice and stop the crash when accessing the enet registers

dleach02 commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -166,6 +177,64 @@ rx_buffer[CONFIG_ETH_MCUX_RX_BUFFERS][ETH_MCUX_BUFFER_SIZE];
+
+			ENET_Deinit(eth_ctx->base);
+			clock_control_off(eth_ctx->clock_dev,
+				(clock_control_subsys_t)eth_ctx->clock);
```
First, add a ENET_Reset() before the Deinit(). I think that will block any transient events that might generate an interrupt.

Second, you don't need the clock_control_off because the ENET_Deinit() already does this. Now @MaureenHelm  and I have been discussing this because one could argue that maybe the NXP HAL shouldn't do that and there are compile options to avoid this but that might have ramifications in other areas.

tbursztyka replied:
Well, I guess atm it does not harm to let the clock_control stuff (it's redundant at worse case).
I'll update the ethernet driver.
